DOC-12667 Server XDCR and mobile coexistence for 7_6_6

rao-shwe opened the pull request:
DOC-12667

rao-shwe committed:
first commit

rao-shwe committed:
added-all-planned-content

rao-shwe committed:
small-changes-in-bookmarks

rao-shwe committed:
draft-1

rao-shwe committed:
diagrams-added

rao-shwe committed:
small-edit

rao-shwe commented:
@hyunjuV and @sumukhbhat2701,

This is a draft document for your review. All authoring work is now complete.

Doc preview link: https://preview.docs-test.couchbase.com/Server-7-6-6/home/index.html

Username: docs
Password:

rao-shwe commented:
**Note to me:**

Pending:
1. Change versionPruningWindowHrs using REST API.
2. Change enableCrossClusterVersioning using REST API and from UI. (Includes screenshots from UI)
3. Interlinking enableCrossClusterVersioning page with other pages.

rao-shwe committed:
edits-page-aliases

rao-shwe committed:
removed-page-aliases-except-for-restapi

rao-shwe committed:
added-back-page-aliases

rao-shwe committed:
screenshots-and-rest-apis

rao-shwe committed:
link-verification

rao-shwe committed:
minor-edits

rao-shwe commented:
Doc preview link: https://preview.docs-test.couchbase.com/Server-7-6-6/home/index.html

Username: docs
Password: kTUj@L&inhk3

sumukhbhat2701 commented on modules/learn/pages/clusters-and-availability/xdcr-conflict-resolution.adoc:
```diff
@@ -1,49 +1,68 @@
+[#eccv-true-in-sequence-number-based-conflict-resolution]
+=== When enableCrossClusterVersioning is true for a sequence number based conflict resolution process 
+
+When the `enableCrossClusterVersioning` property is set to `true`, the hybrid logical vector (HLV) metadata, located in the document xattrs (user created extended attributes) of the source and the target, is also used in the sequence number based conflict resolution processing. The features or options used in the XDCR process determine the specific ways of using the HLV metadata. For more informa ...
```
HLV is a system created extended attribute and not a "user created extended attributes"

sumukhbhat2701 commented on modules/xdcr-reference/pages/xdcr-advanced-settings.adoc:
```diff
@@ -95,6 +95,8 @@ For documents that are smaller, XDCR replicates the document to the target witho
 Note that a low setting risks increased latency, due to a higher number or metadata fetches; but may also reduce the number of required replications (due to source and target having identical copies of the document).
 A high setting lowers latency during replication, since the number of metadata fetches is reduced; but may also raise the replication-rate excessively, overwhelming either network or target cluster.
 
+XDCR optimistic replication operation only applies when the `enableCrossClusterVersioning` property is not being used. When the `enableCrossClusterVersioning` property is set to `true` (enabled), the XDCR processing always retrieves metadata from the target cluster for every document of any size that undergoes replication.
```
The word operation feels odd here. Can we change it to "option" or "concept"

rao-shwe replied:
I've removed that term "operation" and rephrased the sentence a bit.

Changed from:
XDCR optimistic replication operation only applies when the `enableCrossClusterVersioning` property is not being used.

To:
XDCR optimistic replication is applicable only when the `enableCrossClusterVersioning` property is disabled.

sumukhbhat2701 commented on modules/learn/pages/clusters-and-availability/xdcr-conflict-resolution.adoc:
```diff
@@ -57,6 +76,11 @@ If both document-versions have the same timestamp-value, the conflict is resolve
+[#eccv-true-in-timestamp-based-conflict-resolution]
+=== When enableCrossClusterVersioning is true for a timestamp-based conflict resolution process
+
+When the `enableCrossClusterVersioning` property is set to `true`, the hybrid logical vector (HLV) metadata located in the document xattr (user created extended attributes) of the source and the target, is also used in the timestamp-based conflict resolution processing. The features or options used in the XDCR process determine the specific ways of using the HLV metadata. For more information on  ...
```
same here - it should be system created extended attribute.

sumukhbhat2701 commented on modules/learn/pages/clusters-and-availability/xdcr-enable-crossclusterversioning.adoc:
```diff
@@ -0,0 +1,56 @@
+[#manage-version-pruning-window-hrs]
+=== Manage the bucket property versionPruningWindowHrs
+
+You can manage the bucket property `enableCrossClusterVersioning` through REST API.
```
I think this is a copy-paste error wherein the heading says versionPruningWindowHrs and the description is for enableCrossClusterVersioning. Copied from top?

rao-shwe replied:
Yes, indeed, and the effect of working post dinner.

sumukhbhat2701 commented on modules/manage/pages/manage-xdcr/create-xdcr-replication.adoc:
```diff
@@ -214,6 +214,17 @@ If successful, this provides the following response:
+* Create or update a XDCR replication with `mobile=Active` option from UI, starting from Server 7.6.6 version. 
+
+
+The pre-requisite to use `mobile=Active` is to set the bucket property `enableCrossClusterVersioning`. For more information, see xref:learn:clusters-and-availability/xdcr-enable-crossclusterversioning.adoc[XDCR enableCrossClusterVersioning] and xref:learn:clusters-and-availability/conflict-resolution.adoc#the_conflict_resolution_process[The Conflict Resolution Process].
```
The link for The Conflict Resolution Process didn't work for me for some reason in the preview page

rao-shwe replied:
I've fixed it now.

sumukhbhat2701 commented on modules/manage/pages/manage-xdcr/create-xdcr-replication.adoc:
```diff
@@ -214,6 +214,17 @@ If successful, this provides the following response:
+== Create an XDCR Replication with mobile=Active
+
+To create or update an XDCR replication with `mobile=Active`, do the following: 
+* Create a XDCR replication with `mobile=Active` or update an existing replication. For more information, see _Greenfield deployment_ and _Upgrade an existing setup_ in xref:learn:clusters-and-availability/xdcr-active-active-sgw.adoc[XDCR Active-Active with Sync Gateway]
```
The "*" - are they expected to be bullet points? I see them all in the same line.

rao-shwe replied:
Yes, fixed it now.

sumukhbhat2701 commented:
One request, @rao-shwe, can we please remove the `hlvPruningWindowSec` row in https://preview.docs-test.couchbase.com/Server-7-6-6/server/current/rest-api/rest-xdcr-adv-settings.html#xdcr-advanced-settings-rest. It is moved as a bucket setting as we are documenting `versionPruningWindowHrs`. 

In fact we can remove all occurances of `hlvPruningWindowSec` in that page (the examples for instance) to avoid confusion for the reader.

rao-shwe committed:
Updated xdcr-conflict-resolution.adoc

Corrected "xattr (user created extended attributes)" to "xattr (system created extended attributes)"

rao-shwe committed:
Updated xdcr-enable-crossclusterversioning.adoc

rao-shwe committed:
Update xdcr-advanced-settings.adoc

rao-shwe committed:
Updated create-xdcr-replication.adoc

rao-shwe committed:
removed-hlvPruningWindowSec

rao-shwe committed:
minor edit in rest-xdcr-adv-settings.adoc

rao-shwe commented:
> One request, @rao-shwe, can we please remove the `hlvPruningWindowSec` row in https://preview.docs-test.couchbase.com/Server-7-6-6/server/current/rest-api/rest-xdcr-adv-settings.html#xdcr-advanced-settings-rest. It is moved as a bucket setting as we are documenting `versionPruningWindowHrs`.
> 
> In fact we can remove all occurances of `hlvPruningWindowSec` in that page (the examples for instance) to avoid confusion for the reader.

I've removed all instances of hlvPruningWindowSec from the examples and retained hlvPruningWindowSec information in the tabular column.

rao-shwe committed:
minor edit in create-xdcr-replication.adoc

rao-shwe committed:
added link to changes from ui create-xdcr-replication.adoc

rao-shwe committed:
Updated xdcr-enable-crossclusterversioning.adoc

rao-shwe committed:
added_766_release_notes_content

rao-shwe committed:
updated-release-notes-766

rao-shwe committed:
minor edit

rao-shwe committed:
a minor edit

hyunjuV commented on modules/learn/pages/clusters-and-availability/xdcr-enable-crossclusterversioning.adoc:
```diff
@@ -0,0 +1,46 @@
+= XDCR enableCrossClusterVersioning
+:description: pass:q[Hybrid Logical Vector (HLV) must be enabled for each of the buckets participating in the replication process using the `enableCrossClusterVersioning` property. As a result, mobile and XDCR become a part of the same replication process.]
```
Too early to talk about HLV, and the relationship between HLV, Mobile, and XDCR is a bit complicated.  Should just describe the basics here -- like below.

Enabling Cross Cluster Versioning allows XDCR to add metadata to each replicated document.  Enabling this bucket property for all buckets in the replication topology is a pre-requisite for some XDCR features.   The bucket property `enableCrossClusterVersioning` cannot be disabled once it is set to true.  Therefore, you should understand all the pre-requisites for a feature, and the full effects of setting this bucket property to true, before enabling it.

hyunjuV commented on modules/learn/pages/clusters-and-availability/xdcr-enable-crossclusterversioning.adoc:
```diff
@@ -0,0 +1,46 @@
+{description}
+
+
+When you set the bucket property `enableCrossClusterVersioning` (ECCV) to `true`, for each document processed by XDCR, XDCR stores an internal history of the document in the extended attributes. This set of information is also called Hybrid Logical Vector (HLV), which is a set of Hybrid Logical Clock (HLC) information.
```
I know that I used the "internal history of the document" phrasing, but I've changed my mind on this -- we should just say that additional metadata is stored.

When you set the bucket property `enableCrossClusterVersioning` (ECCV) to `true`, for each document processed by XDCR, XDCR stores additional metadata for the document in the extended attributes.  This metadata is also called the Hybrid Logical Vector (HLV), which is a set of Hybrid Logical Clock (HLC) information.

hyunjuV commented on modules/learn/pages/clusters-and-availability/xdcr-enable-crossclusterversioning.adoc:
```diff
@@ -0,0 +1,46 @@
+
+When you set the bucket property `enableCrossClusterVersioning` (ECCV) to `true`, for each document processed by XDCR, XDCR stores an internal history of the document in the extended attributes. This set of information is also called Hybrid Logical Vector (HLV), which is a set of Hybrid Logical Clock (HLC) information.
+
+NOTE: The bucket property `enableCrossClusterVersioning` must be set to `true` to use the `mobile=[Off | Active]` flag, and this flag in turn  enables the setting _XDCR Active-Active with Sync Gateway 4.0+_. For more information about XDCR Active-Active with Sync Gateway 4.0+, see xref:learn:clusters-and-availability/xdcr-active-active-sgw.adoc[XDCR Active-Active with Sync Gateway].
```
1. This Note might be confusing for a person who doesn't know where the mobile flag/option is used -- so, adding the info that it is a flag/option in create/update replication.
2. This explanation/note should be in a separate section called "Dependent XDCR features" and explain in that section that:

The bucket property `enableCrossClusterVersioning` must be set to `true` to use the `mobile=[Off | Active]` flag while creating or updating a replication.  The mobile flag must be set to Active for bi-directional XDCR with Sync Gateway 4.0+.  For more information about XDCR Active-Active with Sync Gateway 4.0+, see xref:learn:clusters-and-availability/xdcr-active-active-sgw.adoc[XDCR Active-Active with Sync Gateway].

hyunjuV commented on modules/learn/pages/clusters-and-availability/xdcr-enable-crossclusterversioning.adoc:
```diff
@@ -0,0 +1,46 @@
+
+The HLV metadata takes up `109 + 40N` bytes of space per document, where N is the number of buckets mutating the document in the replication topology. 
+
+NOTE: The size of the HLV metadata must remain constant as long as your replication topology is constant. However, the HLV data accumulation occurs when the replication topology changes and the document copy goes through different clusters, for example, when being restored to a new cluster.
```
The size of the HLV metadata can be less than `109 + 40N` -- it'll grow to that limit as long as the replication topology does not change.   So, change to:

NOTE:  As long as your replication topology is constant, the size of the HLV metadata will grow to `109 + 40N` and remain constant.  However, the HLV data accumulation occurs when the replication topology changes and the document copy goes through different clusters, for example, when being restored to a new cluster.

hyunjuV commented on modules/learn/pages/clusters-and-availability/xdcr-enable-crossclusterversioning.adoc:
```diff
@@ -0,0 +1,46 @@
+
+To remove the accumulated or outdated HLV data, the HLV metadata is pruned periodically. You can control the pruning frequency by setting the bucket property `versionPruningWindowHrs`.
+
+IMPORTANT: This is an important limitation to the _XDCR Active-Active with Sync Gateway_ feature. If you use the user created extended attributes (xattrs) in your documents, and you have more than 10 user xattrs in a document, then you cannot use the feature _XDCR Active-Active with Sync Gateway_. This is due to an internal limitation of managing extended attributes in a document. If you try to u ...
```
This IMPORTANT note should be in the section "Dependent XDCR features" under the paragraph about the mobile setting.

Currently, the mobile/XDCR feature information is mixed up with the HLV information.   So, the mobile/XDCR information should be moved to its own section called "Dependent XDCR features".

hyunjuV commented on modules/learn/pages/clusters-and-availability/xdcr-enable-crossclusterversioning.adoc:
```diff
@@ -0,0 +1,46 @@
+
+You can manage the bucket property `versionPruningWindowHrs` through REST API as follows:
+
+* For information about creating the bucket property `versionPruningWindowHrs` through REST API, see xref:rest-api:rest-bucket-create.adoc#curl-syntax[Curl Syntax].
```
You don't need this info about creating the bucket property `versionPruningWindowHrs` -- it's always exists when a bucket is created.  So, remove this sentence.

hyunjuV commented on modules/learn/pages/clusters-and-availability/xdcr-enable-crossclusterversioning.adoc:
1. There should be a section called "Dependent XDCR features". Currently, the mobile/XDCR feature information is mixed up with the HLV information. So, the mobile/XDCR information should be moved to its own section called "Dependent XDCR features".  I've noted in the comments which paragraphs/notes should be moved into that section.

Also, the following additional information should be provided:

Bi-directional replication with  Sync Gateway  and XDCR
The other pre-requisite for bi-directional replication with Sync Gateway and XDCR is to use Sync Gateway version 4.0 and above.  Please note that if you are using Sync Gateway version 4.0 and above, Sync Gateway itself will add HLV info (regardless of whether you have actively set enableCrossClusterVersioning to true in the bucket property).  The existence of the HLV info will cause XDCR to maintain that info, so for one-way (unidirectional) XDCR/Sync Gateway users, just using Sync Gateway version 4.0 and above will mean that there are additional metadata.  You still need to enable the bucket property enableCrossClusterVersioning for all buckets in the replication topology and set “mobile” to “Active” in the XDCR replication to configure bi-directional replication with Sync Gateway and XDCR.

2.  There should be a section on "Removing HLV metadata" and explain:
After enabling, you cannot disable the Cross Cluster Versioning bucket property.  You can, however, backup and restore the data to a bucket with "enableCrossClusterVersioning" set to false and use the `cbbackupmgr restore` `--disable-hlv` option to strip the xattr info added by XDCR.  To remove the on-going maintenance of HLV info, you must remove the xattr info previously added to the documents.  Just restoring the data to a bucket where enableCrossClusterVersioning is set to false is not enough to completely remove the effects of having enabled HLV.

hyunjuV reviewed (changes requested):
This is just the review of the new page "XDCR enableCrossClusterVersioning".

rao-shwe committed:
review-comments-of-pm-implemented

rao-shwe committed:
Update new_features-76_6.adoc

rao-shwe committed:
minor edit in xdcr-enable-crossclusterversioning.adoc

rao-shwe committed:
Update rest-bucket-create.adoc

hyunjuV commented on modules/learn/pages/clusters-and-availability/xdcr-active-active-sgw.adoc:
```diff
@@ -0,0 +1,79 @@
+= XDCR Active-Active with Sync Gateway
+:description: pass:q[Using XDCR Active-Active replication with Sync Gateway, you can configure an active-active XDCR setup with both Sync Gateway (SGW) and mobile applications on the XDCR source and target clusters.]
```
Because of the odd circumstances of this full feature availability, I think that we should strike a different tone here -- more towards making sure that an inadvertent mistake isn't made.  I realize that you pretty much explain the same thing in the Introduction, but I still think that it's worth repeating because people don't always read carefully or everything.

Suggestion:
You can use XDCR with Sync Gateway mobile clusters in a bi-directional, active - active replication, but you must make sure that both the Server and the Sync Gateway versions support this option.  Otherwise, using XDCR with Sync Gateway buckets in a bi-directional replication can cause data corruption.

hyunjuV commented on modules/learn/pages/clusters-and-availability/xdcr-active-active-sgw.adoc:
```diff
@@ -0,0 +1,79 @@
+Here are a few limitations to the XDCR active-active with Sync Gateway feature.
+
+* If you use the user created extended attributes (xattrs) in your documents, and you have more than 10 user xattrs in a document, then you cannot use the feature _XDCR Active-Active with Sync Gateway_. This is due to an internal limitation of managing extended attributes in a document. If you try to use the feature _XDCR Active-Active with Sync Gateway_ when you have more than 10 user xattrs in  ...
+* Eventing Service cannot yet be used in a Sync Gateway 4.0+ version or in an XDCR bi-directional replication environment because a metadata update occurs causing XDCR to ping-pong and never stop replicating in an Active-Active XDCR environment. However, one-way replication is possible.
```
The phrasing for the Eventing Service information is a bit confusing.  Since we don't know when the fix to this issue will be done, we should just say that they cannot use Eventing with Sync Gateway and bi-directional XDCR.  (Once the fix is available, we can update the documentation for the Server version where the fix is available.)

* Eventing Service cannot be used with Sync Gateway and bi-directional XDCR.  If used with Sync Gateway in a bi-directional, active-active XDCR environment, the Eventing Service metadata updates in the source and target clusters will cause XDCR to ping-pong and never stop replicating.  One-way replication is possible.

hyunjuV commented on modules/learn/pages/clusters-and-availability/xdcr-active-active-sgw.adoc:
```diff
@@ -0,0 +1,79 @@
+[#xdcr-active-active-sgw-prerequisites]
+== Prerequisites
+
+Set the bucket property `enableCrossClusterVersioning` to use the setting `mobile=Active`. To enable the bucket property `enableCrossClusterVersioning` using REST API or from the UI, see xref:learn:clusters-and-availability/xdcr-enable-crossclusterversioning.adoc[XDCR enableCrossClusterVersioning].
```
The first sentence should include the phrase "when creating or updating the replication settings" (like below) since people won't know where the setting mobile=Active is used:

Set the bucket property `enableCrossClusterVersioning` to use the setting `mobile=Active` when creating or updating the replication settings.

rao-shwe replied:
I changed it to similar as in enableCrossClusterVersioning page:

Set the bucket property `enableCrossClusterVersioning` to use the setting `mobile=Active` during the processes [Create a Replication](https://docs.couchbase.com/server/current/manage/manage-xdcr/create-xdcr-replication.html) and [Manage XDCR](https://docs.couchbase.com/server/current/manage/manage-xdcr/xdcr-management-overview.html).

hyunjuV commented on modules/learn/pages/clusters-and-availability/xdcr-active-active-sgw.adoc:
```diff
@@ -0,0 +1,79 @@
+NOTE: ECCV refers to the bucket property `enableCrossClusterVersioning`. 
++
+. Enable ECCV on bucket B2. All the mutations in B2, after this point of time, will have a new metadata called HLV.
+. Update the replication settings to `mobile=Active` of the already existing XDCR from B1 to B2.
```
Should add here:

You can use the REST API or the XDCR UI to update an existing replication.  For information on using the REST API to modify replication settings for an existing replication, see Managing Advanced Settings -- https://docs.couchbase.com/server/current/rest-api/rest-xdcr-adv-settings.html

Also, it would be nice to have an example REST API of updating a replication and setting mobile=Active since there isn't one elsewhere.  Maybe it can be put in the Managing Advanced Settings page or in this page some where.
 
Example of updating an existing replication to set mobile = Active using the REST API:
curl -X POST -u Administrator:password \
http://localhost:8091/settings/replications/409b66f64b65193352acc3354e034fac%2Ftestbucket%2Ftestbucket -d mobile=Active

The below two info is already in the Managing Advanced Settings page:
Example of the REST API call to view the replication settings:
curl -X GET -u Administrator:password \
http://localhost:8091/settings/replications/409b66f64b65193352acc3354e034fac%2Ftestbucket%2Ftestbucket

The replication settingsURI (the replication id) can be retrieved from the GET /pools/default/tasks REST API call, as explained in the XDCR API Managing Advanced Settings documentation.

hyunjuV reviewed (changes requested):
These comments are for the review of the new page XDCR Active-Active with Sync Gateway.

hyunjuV commented on modules/introduction/partials/new_features-76_6.adoc:
```diff
@@ -0,0 +1,7 @@
+=== XDCR
+
+* *https://jira.issues.couchbase.com/browse/MB-57921[MB-57921]:*
+Activated the usage of an already existing bucket property `enableCrossClusterVersioning` (ECCV). When `enableCrossClusterVersioning` is set to `true`, for each document processed by XDCR, XDCR stores additional metadata for the document in the extended attributes. This metadata is also called Hybrid Logical Vector (HLV), which is a set of Hybrid Logical Clock (HLC) information. For more informat ...
```
This paragraph should be removed since ECCV is not a feature in itself.  It's a pre-requisite.  We don't really want people thinking that ECCV is a feature in itself and accidentally enabling it.  We really only want people to go look at the ECCV info because enabling it is a requirement for a feature that they do want to use.

hyunjuV commented on modules/introduction/partials/new_features-76_6.adoc:
```diff
@@ -0,0 +1,7 @@
+* *https://jira.issues.couchbase.com/browse/MB-57921[MB-57921]:*
+Activated the usage of an already existing bucket property `enableCrossClusterVersioning` (ECCV). When `enableCrossClusterVersioning` is set to `true`, for each document processed by XDCR, XDCR stores additional metadata for the document in the extended attributes. This metadata is also called Hybrid Logical Vector (HLV), which is a set of Hybrid Logical Clock (HLC) information. For more informat ...
++
+Created provision to set up XDCR bi-directional replication with Sync Gateway 4.0+. You can configure an active-active XDCR setup with both Sync Gateway (SGW) and mobile applications on the XDCR source and target clusters. The bucket property `enableCrossClusterVersioning` must be set to `true` to use the `mobile=[Off | Active]` flag. Then the `mobile=[Off | Active]` flag must be set to `Active`  ...
```
Since you've provided a great doc page on XDCR Active-Active with Sync Gateway already, I don't think that you should go into so much detail here -- the details are confusing here.  Here, should just say what you said in the xdcr/sgw doc page intro, since it's a great summary:

Created provision to set up XDCR bi-directional replication with Sync Gateway 4.0+.  In the versions earlier than Server 7.6.6 and Sync Gateway (SGW) 4.0.0, only an active-passive setup was supported with both XDCR and SGW. XDCR Active-Active replication with Sync Gateway for XDCR-Mobile interoperability configuration is introduced in the Server 7.6.6 version, where you can configure an active-active XDCR setup with Sync Gateway and mobile applications both on the XDCR source and target clusters. You need to have at least a Server 7.6.6 version and SGW 4.0.0 version to use this setup.  For more info, see xref:learn:clusters-and-availability/xdcr-active-active-sgw.adoc[XDCR Active-Active with Sync Gateway].

hyunjuV reviewed (changes requested):
Review of What's New summary.

hyunjuV commented on modules/learn/pages/clusters-and-availability/xdcr-conflict-resolution.adoc:
```diff
@@ -1,49 +1,68 @@
 = XDCR Conflict Resolution
-:description: pass:q[_XDCR Conflict Resolution_ automatically synchronizes document-copies that have been modified in different ways at different locations.]
+:description: pass:q[In an XDCR process, modified documents are copied or replicated from the source bucket (or collection) to the target bucket (or collection).]
```
I realize that I'm the one who had suggested this change, but now I've decided that it would be better to keep the original description here.  So, please remove this change.

hyunjuV commented on modules/learn/pages/clusters-and-availability/xdcr-conflict-resolution.adoc:
```diff
@@ -1,49 +1,68 @@
 
 [abstract]
 {description}
+If a target document with the same document ID as the source document already exists, the XDCR conflict resolution process determines whether the source document replaces the target document or not.
```
Same with this sentence -- changed my mind -- please remove this new description and just keep the description that was here before (which is the sentence below):

_XDCR Conflict Resolution_ automatically synchronizes document-copies that have been modified in different ways at different locations.

rao-shwe replied:
I'll not mention this because it will be the same sentence mentioned consecutively on the doc page:

_XDCR Conflict Resolution_ automatically synchronizes document-copies that have been modified in different ways at different locations. 

Instead, I'll remove the sentence: 

If a target document with the same document ID as the source document already exists, the XDCR conflict resolution process determines whether the source document replaces the target document or not.

hyunjuV commented on modules/learn/pages/clusters-and-availability/xdcr-conflict-resolution.adoc:
```diff
@@ -1,49 +1,68 @@
+* Timestamp-based conflict resolution.
 
 [#the_conflict_resolution_process]
 == The Conflict Resolution Process
```
The entire "The Conflict Resolution Process" should be removed.
We are trying to describe our internal algorithms, which are changing, so, we shouldn't even try.

rao-shwe replied:
I've removed the complete section "The Conflict Resolution Process".

hyunjuV commented on modules/learn/pages/clusters-and-availability/xdcr-conflict-resolution.adoc:
```diff
@@ -1,49 +1,68 @@
-For documents above 256 bytes in size, XDCR fetches metadata from the target cluster before replicating.
-The target metadata for the document is compared with the source metadata for the document, in order to choose which document should prevail (the exact subset of metadata used in this comparison depends on the source bucket's _conflict resolution policy_).
-If the source document prevails, it is replicated to the target; if the target document prevails, the source document is not replicated.
+The conflict resolution processes can be enabled or disabled using the `enableCrossClusterVersioning` property (ECCV property).
```
Remove.

hyunjuV commented on modules/learn/pages/clusters-and-availability/xdcr-conflict-resolution.adoc:
```diff
@@ -1,49 +1,68 @@
 
-Once a replicated document reaches the target, the target cluster also performs a metadata comparison as described, in order to confirm that the document from the source cluster should indeed prevail. If this is confirmed, the document from the source cluster is applied to the target cluster, and the target cluster's previous version of the document is discarded.
+[#eccv-false-for-conflict-resolution-process]
+=== When enableCrossClusterVersioning is false for a Conflict Resolution Process
```
Remove.

hyunjuV commented on modules/learn/pages/clusters-and-availability/xdcr-conflict-resolution.adoc:
```diff
@@ -1,49 +1,68 @@
+=== When enableCrossClusterVersioning is false for a Conflict Resolution Process
 
-As a performance optimization, XDCR makes no metadata comparison on the source for documents of 256 bytes or less, thus making unnecessary a metadata fetch from the target cluster: instead, the document is replicated immediately to the target, and metadata comparison is performed there.
+When a source document is modified, XDCR determines whether this document revision must be applied to the target document. For documents above the size of 256 bytes, before the replication process, XDCR fetches metadata from the target cluster. To choose the document that must prevail (the exact subset of metadata used in this comparison depends on the conflict resolution policy of source bucket) ...
```
Remove.

hyunjuV commented on modules/learn/pages/clusters-and-availability/xdcr-conflict-resolution.adoc:
```diff
@@ -1,49 +1,68 @@
+When a source document is modified, XDCR determines whether this document revision must be applied to the target document. For documents above the size of 256 bytes, before the replication process, XDCR fetches metadata from the target cluster. To choose the document that must prevail (the exact subset of metadata used in this comparison depends on the conflict resolution policy of source bucket) ...
 
-If a document is deleted on the source, XDCR makes no metadata comparison on the source before replication.
+After the replicated document reaches the target, the target cluster also performs a metadata comparison, as described earlier, to confirm if the document from the source cluster must prevail or not. If it is confirmed that the document from the source cluster must prevail, then the document from the source cluster is applied to the target cluster, and the previous version of the target cluster d ...
```
Remove.

hyunjuV commented on modules/learn/pages/clusters-and-availability/xdcr-conflict-resolution.adoc:
```diff
@@ -1,49 +1,68 @@
+
+XDCR optimistic replication behavior, as described earlier, applies only when the 'enableCrossClusterVersioning' property is not in use. 
+
+When the `enableCrossClusterVersioning` property is set to `true` (enabled), the hybrid logical vector (HLV) metadata is created and maintained for each document in the document xattrs (system created extended attributes). This HLV metadata needs to be checked and updated for both the source and the target documents. As a result, the XDCR processing retrieves metadata from the target cluster for  ...
```
Remove lines 29-46

hyunjuV commented on modules/learn/pages/clusters-and-availability/xdcr-conflict-resolution.adoc:
```diff
@@ -1,49 +1,68 @@
-A _conflict_ is caused when the source and target copies of an XDCR-replicated document are updated independently of and dissimilarly to one another, each by a local application.
-The conflict must be _resolved_, by determining which of the variants should prevail; and then correspondingly saving both documents in identical form.
-XDCR provides an automated _conflict resolution_ process.
+A conflict occurs when a document in the target bucket (or collection) has the same document ID as that of the source bucket (or collection) during an XDCR process. XDCR provides an automated conflict resolution process. XDCR supports the following two alternative conflict resolution policies:
```
Change this line 12 to:

When a source document is modified, XDCR determines whether this revision of the document should be applied to the target.  This process is called conflict resolution, and it is a fully automated process. XDCR supports the following two alternative conflict resolution policies:

hyunjuV commented on modules/learn/pages/clusters-and-availability/xdcr-conflict-resolution.adoc:
```diff
@@ -1,49 +1,68 @@
 
-Two, alternative conflict resolution policies are supported: _sequence-number-based_ (which is the default), and _timestamp-based_.
-Note that _timestamp-based_ conflict resolution is only available in the Enterprise Edition of Couchbase Server.
+* Sequence number-based conflict resolution (This is the default policy). 
+* Timestamp-based conflict resolution.
```
Remove lines 14-15.

In this section, just explain:

When a source document is modified, XDCR determines whether this revision of the document should be applied to the target.  This process is called conflict resolution, and it is a fully automated process. Two, alternative conflict resolution policies are supported: sequence-number-based (which is the default), and timestamp-based.

hyunjuV commented on modules/learn/pages/clusters-and-availability/xdcr-conflict-resolution.adoc:
```diff
@@ -1,49 +1,68 @@
+If a target document with the same document ID as the source document already exists, the XDCR conflict resolution process determines whether the source document replaces the target document or not.
 
 [#conflicts_and_their_resolution]
 == Conflicts and Their Resolution
```
Can we rename this section to "Conflict Resolution" instead of being called "Conflicts and Their Resolution"?

hyunjuV commented on modules/learn/pages/clusters-and-availability/xdcr-conflict-resolution.adoc:
```diff
@@ -57,6 +76,11 @@ If both document-versions have the same timestamp-value, the conflict is resolve
 . Expiration (TTL) value
 . Document flags
 
+[#eccv-true-in-timestamp-based-conflict-resolution]
+=== When enableCrossClusterVersioning is true for a timestamp-based conflict resolution process
+
+When the `enableCrossClusterVersioning` property is set to `true`, the hybrid logical vector (HLV) metadata located in the document xattr (system created extended attributes) of the source and the target, is also used in the timestamp-based conflict resolution processing. The features or options used in the XDCR process determine the specific ways of using the HLV metadata. For more information o ...
```
Remove lines 79-82.

Add a line that says:
When Cross Cluster Versioning is enabled, the hybrid logical vector (HLV) metadata in the source and target document's xattr, is also used in the conflict resolution processing.  For more information on the `enableCrossClusterVersioning` property and the HLV metadata, see xref:clusters-and-availability/xdcr-enable-crossclusterversioning.adoc[XDCR enableCrossClusterVersioning].

hyunjuV reviewed (changes requested):
These are the review comments for the conflict resolution page.

hyunjuV commented on modules/learn/pages/clusters-and-availability/xdcr-conflict-resolution.adoc:
```diff
@@ -1,49 +1,68 @@
 . Expiration (TTL) value
 . Document flags
 
+[#eccv-true-in-sequence-number-based-conflict-resolution]
+=== When enableCrossClusterVersioning is true for a sequence number based conflict resolution process 
+
+When the `enableCrossClusterVersioning` property is set to `true`, the hybrid logical vector (HLV) metadata, located in the document xattrs (system created extended attributes) of the source and the target, is also used in the sequence number based conflict resolution processing. The features or options used in the XDCR process determine the specific ways of using the HLV metadata. For more infor ...
```
Remove lines 61-64.
Add a line that says:
When Cross Cluster Versioning is enabled, the hybrid logical vector (HLV) metadata in the source and target document's xattr, is also used in the conflict resolution processing. For more information on the enableCrossClusterVersioning property and the HLV metadata, see xref:clusters-and-availability/xdcr-enable-crossclusterversioning.adoc[XDCR enableCrossClusterVersioning].

hyunjuV commented on modules/learn/pages/clusters-and-availability/xdcr-enable-crossclusterversioning.adoc:
```diff
@@ -0,0 +1,65 @@
+{description}
+
+
+Enabling the bucket property `enableCrossClusterVersioning` for all buckets in the replication topology is a pre-requisite for some XDCR features. The bucket property `enableCrossClusterVersioning` cannot be disabled once it is set to `true`. Therefore, you must understand all the pre-requisites and the full effects of setting `enableCrossClusterVersioning` to `true` before enabling it.
```
Looking at this page, this paragraph seemed to have too much repetition.  Maybe reword it a bit.  Something like:

Enabling Cross Cluster Versioning for all buckets in the replication topology is a pre-requisite for some XDCR features.  The bucket property `enableCrossClusterVersioning` cannot be disabled once it is set to `true`. Therefore, you should not enable frivolously.

hyunjuV commented on modules/learn/pages/clusters-and-availability/xdcr-enable-crossclusterversioning.adoc:
```diff
@@ -0,0 +1,65 @@
+When you set the bucket property `enableCrossClusterVersioning` (ECCV) to `true`, for each document processed by XDCR, XDCR stores additional metadata for the document in the extended attributes. This metadata is also called Hybrid Logical Vector (HLV), which is a set of Hybrid Logical Clock (HLC) information.
+
+[#dependent-xdcr-features]
+== Dependent XDCR features
```
This section should be moved to the end of the page.  This way, the above mention of HLV can flow into the section called "Hybrid Logical Vector (HLV) data maintained in the extended attributes".

Also, instead of calling it "Dependent XDCR features", let's call it "Features that require enabling Cross Cluster Versioning", and we can just list the Sync Gateway/XDCR feature with a link to the Sync Gateway/XDCR page that you created.

hyunjuV commented on modules/learn/pages/clusters-and-availability/xdcr-enable-crossclusterversioning.adoc:
```diff
@@ -0,0 +1,65 @@
+[#dependent-xdcr-features]
+== Dependent XDCR features
+
+This section lists feature settings on which replication related XDCR-mobile interoperability depends.
```
Please remove this sentence.  Now that I'm looking at this page, I don't think that we should repeat the same information that you already have in the "XDCR Active-Active with Sync Gateway" page.

We should say:
The bi-directional, active-active replication with Sync Gateway 4.0+ and XDCR requires enabling Cross Cluster Versioning.  For more information, including important limitations, see xref:learn:clusters-and-availability/xdcr-active-active-sgw.adoc[XDCR Active-Active with Sync Gateway].

hyunjuV commented on modules/learn/pages/clusters-and-availability/xdcr-enable-crossclusterversioning.adoc:
```diff
@@ -0,0 +1,65 @@
+
+This section lists feature settings on which replication related XDCR-mobile interoperability depends.
+
+[#bidirectional-replication-with-sgw-and-xdcr]
+=== Bi-directional replication with Sync Gateway and XDCR
+
+Pre-requisites for a bi-directional replication with Sync Gateway and XDCR are as follows:
+
+* Set the bucket property `enableCrossClusterVersioning` to `true` to use the `mobile=[Off | Active]` flag during the processes xref:manage:manage-xdcr/create-xdcr-replication.adoc[Create a Replication] and xref:manage:manage-xdcr/xdcr-management-overview.adoc[Manage XDCR]. Then set the `mobile=[Off | Active]` flag to `Active` for a bi-directional replication with Sync Gateway 4.0+ and XDCR.
+
+* Use Sync Gateway 4.0 or a higher version. When you're using Sync Gateway 4.0 or a higher version, Sync Gateway adds HLV information (whether or not you have actively set `enableCrossClusterVersioning` to `true` in the bucket property). The existence of the HLV information causes XDCR to maintain that information, so for one-way (unidirectional) XDCR-Sync Gateway users, using Sync Gateway 4.0 or ...
+
+For information about XDCR Active-Active with Sync Gateway 4.0+ (or to configure bi-directional replication with Sync Gateway and XDCR), see xref:learn:clusters-and-availability/xdcr-active-active-sgw.adoc[XDCR Active-Active with Sync Gateway].
+
+IMPORTANT: This is an important limitation to the _XDCR Active-Active with Sync Gateway_ feature. If you use the user created extended attributes (xattrs) in your documents, and you have more than 10 user xattrs in a document, then you cannot use the feature _XDCR Active-Active with Sync Gateway_. This is due to an internal limitation of managing extended attributes in a document. If you try to u ...
```
Should remove this entire section -- lines 17 to 28.

hyunjuV commented on modules/learn/pages/clusters-and-availability/xdcr-enable-crossclusterversioning.adoc:
```diff
@@ -0,0 +1,65 @@
+[#hlv-data-maintained-in-xattr]
+== Hybrid Logical Vector (HLV) data maintained in the extended attributes
+
+The new metadata, HLV, is stored as a system user created extended attribute (xattr). The HLV metadata is also called Version Vectors. The system xattr name is `"_vv" -- "xattrs"."_vv"`. 
```
The first sentence is a bit confusing -- it says "system user created" -- I think it should just say "system created" (otherwise, you might think, is it "system created" or "user created").

The new metadata, HLV, is stored as a system created extended attribute (xattr) called `_vv` (`xattrs._vv`).  The HLV metadata is also called Version Vectors.

hyunjuV commented on modules/learn/pages/clusters-and-availability/xdcr-enable-crossclusterversioning.adoc:
```diff
@@ -0,0 +1,65 @@
+= XDCR enableCrossClusterVersioning
+:description: pass:q[Enabling the bucket property `enableCrossClusterVersioning` allows XDCR to add metadata to each replicated document.]
```
Since we use `enableCrossClusterVersioning` in the next paragrah, as well as explaining that it's a bucket property, here, instead of:
Enabling the bucket property `enableCrossClusterVersioning` allows XDCR to add metadata to each replicated document.
Let's say:
Enabling Cross Cluster Versioning allows XDCR to add metadata to each replicated document.

hyunjuV reviewed (changes requested):
Some additional changes to the XDCR ECCV page.  Saw the changes you had made -- they looked good, but I thought that we should rename one of the sections and move it to the bottom of the page so that the information flowed better.  Also, looking at the page and the information you already have in the Sync Gateway/XDCR page, I realized that we don't need to repeat it on this page.

hyunjuV commented on modules/manage/pages/manage-xdcr/create-xdcr-replication.adoc:
```diff
@@ -214,6 +214,17 @@ If successful, this provides the following response:
+* Create or update a XDCR replication with `mobile=Active` option using the REST API, starting from Server 7.6.6 version. See xref:rest-api/rest-xdcr-create-replication.adoc[Create a Replication].
+* Create or update a XDCR replication with `mobile=Active` option from UI, starting from Server 7.6.6 version. See xref:manage:manage-xdcr/create-xdcr-replication.adoc#create-an-xdcr-replication-with-the-ui[Create an XDCR Replication with the UI].
+
+The pre-requisite to use `mobile=Active` is to set the bucket property `enableCrossClusterVersioning`. For more information, see xref:learn:clusters-and-availability/xdcr-enable-crossclusterversioning.adoc[XDCR enableCrossClusterVersioning] and xref:learn:clusters-and-availability/xdcr-conflict-resolution.adoc#the_conflict_resolution_process[The Conflict Resolution Process].
```
In the above sentence, remove the reference to The Conflict Resolution Process since that section should be removed.   Should just say:

The pre-requisite to use `mobile=Active` is to set the bucket property `enableCrossClusterVersioning`. For more information, see xref:learn:clusters-and-availability/xdcr-enable-crossclusterversioning.adoc[XDCR enableCrossClusterVersioning].

rao-shwe replied:
Thanks for mentioning this instance.

hyunjuV reviewed (changes requested):
Review of create xdcr replication page.

hyunjuV commented on modules/learn/pages/clusters-and-availability/xdcr-enable-crossclusterversioning.adoc:
```diff
@@ -0,0 +1,65 @@
+== Modify the bucket property enableCrossClusterVersioning
+
+You can modify the bucket property `enableCrossClusterVersioning` through REST API. For information about modifying `enableCrossClusterVersioning` through REST API, see xref:rest-api:rest-bucket-create.adoc#example-enablecrossclusterversioning-edit[Example: Turning on enableCrossClusterVersioning, when Editing].
+
```
Note:  You cannot enable Cross Cluster Versioning while creating the bucket.

hyunjuV commented on modules/rest-api/pages/rest-bucket-create.adoc:
```diff
@@ -1090,6 +1091,46 @@ This parameter is ignored if `autoCompactionDefined` is `false` (which is its de
+[#enablecrossclusterversioning]
+=== enableCrossClusterVersioning
+
+When you change the bucket property `enableCrossClusterVersioning` to `true`, for each document processed by XDCR, XDCR stores additional metadata for the document in the extended attributes. This set of information is also called Hybrid Logical Vector (HLV), which is a set of Hybrid Logical Clock (HLC) information. HLV must be enabled for each of the buckets participating in the replication proc ...
```
Please remove this sentence:
"So, mobile and XDCR are a part of the same replication proccess."

Please add the info that enabling this bucket property is a pre-requisite to some XDCR features.  We don't need to say which ones here.  When the users go to the XDCR ECCV page, they'll find out. (Or, the only reason they'll want to know anything about this bucket property is because a feature that they want to use says it's a pre-requisite.)

Something like this:
Enabling Cross Cluster Versioning is a pre-requisite to some XDCR features.  The bucket property `enableCrossClusterVersioning` can only be set to `true` after a bucket has been created.  When enabled, for each document processed by XDCR, XDCR stores additional metadata, called the Hybrid Logical Vector (HLV), in the document extended attributes (xattrs). For more information, see xref:clusters-and-availability/xdcr-enable-crossclusterversioning.adoc[XDCR enableCrossClusterVersioning].

hyunjuV commented on modules/rest-api/pages/rest-bucket-create.adoc:
```diff
@@ -1090,6 +1091,46 @@ This parameter is ignored if `autoCompactionDefined` is `false` (which is its de
+
+See the example provided in xref:rest-api:rest-bucket-create.adoc#example-enablecrossclusterversioning-edit[Example: Turning on enableCrossClusterVersioning, when Editing]
+
+CAUTION: The default value is `false`. Do not change the value of this property. Once enabled, you cannot turn off the `enableCrossClusterVersioning` property. The only way for you to undo setting this value to `true` is to backup your data, create a new bucket, and reload the data into it.
```
CAUTION: The default value is `false`.  Do not change the value of this property unless instructed by a feature configuration.  Once enabled, you cannot turn off the `enableCrossClusterVersioning` property. The only way for you to undo setting this value to `true` is to backup your data, create a new bucket, and restore the data, using the `cbbackupmgr restore` option `--disable-hlv` to strip the HLV info in the xattrs.

rao-shwe replied:
I assume, it is (and mentioned as follows in the doc). Let me know if the following is incorrect. I referred to this existing page (https://docs.couchbase.com/server/current/backup-restore/cbbackupmgr-restore.html#synopsis):

... using the option `cbbackupmgr restore [--disable-hlv]`

rao-shwe replied:
@hyunjuV 

Should I list `cbbackupmgr restore [--disable-hlv]` on this page (https://docs.couchbase.com/server/current/backup-restore/cbbackupmgr-restore.html); in the sections [Synopsis](https://docs.couchbase.com/server/current/backup-restore/cbbackupmgr-restore.html#synopsis) and [Optional](https://docs.couchbase.com/server/current/backup-restore/cbbackupmgr-restore.html#optional) ?

hyunjuV replied:
@rao-shwe 
Yes.  Thanks for catching that it needs to be documented.
I've created DOC-13240 -- https://jira.issues.couchbase.com/browse/DOC-13240

hyunjuV commented on modules/xdcr-reference/pages/xdcr-advanced-settings.adoc:
```diff
@@ -95,6 +95,8 @@ For documents that are smaller, XDCR replicates the document to the target witho
 Note that a low setting risks increased latency, due to a higher number or metadata fetches; but may also reduce the number of required replications (due to source and target having identical copies of the document).
 A high setting lowers latency during replication, since the number of metadata fetches is reduced; but may also raise the replication-rate excessively, overwhelming either network or target cluster.
 
+XDCR optimistic replication is applicable only when the `enableCrossClusterVersioning` property is disabled. When the `enableCrossClusterVersioning` property is set to `true` (enabled), the XDCR processing always retrieves metadata from the target cluster for every document of any size that undergoes replication.
```
Part of this description is no longer true -- at least, it depends on the scenario since we've introduced other optimizations using HLV.  So, for simplicity, we should just say (just keep the first sentence and remove the rest):

XDCR optimistic replication is applicable only when the `enableCrossClusterVersioning` property is disabled.

hyunjuV reviewed (changes requested):
Reviewed the rest of the changes.  Mostly looked great, just a few changes requested.

hyunjuV commented on modules/rest-api/pages/rest-xdcr-adv-settings.adoc:
```diff
@@ -455,6 +452,11 @@ This setting can be established and retrieved either for an individual replicati
+| Active or Off
+| Default: `Off`.
+When set to `Active`, enables the setting _XDCR Active-Active with Sync Gateway 4.0+_ on the clusters of both sides of the replication. The default value `Off` indicates that the replication setup supports either _XDCR Active-Passive with Sync Gateway_ or _XDCR Active-Active without Sync Gateway_. For more information, see xref:learn:clusters-and-availability/xdcr-active-active-sgw.adoc[XDCR Acti ...
+
```
This `mobile` entry is fine.

A few lines down -- lines 469 to 474, there is the `optimisticReplicationThreshold`.

For the optimisticReplicationThreshold, we should add the below sentence like you did for the optimistic replication info in the advanced settings page:

XDCR optimistic replication is applicable only when the `enableCrossClusterVersioning` property is disabled.

rao-shwe commented:
Note to me:
Verify if "The Conflict Resolution Process" section is referred in any another doc page and remove those reference links.

rao-shwe committed:
implemented-pm-review-comments

hyunjuV commented on modules/learn/pages/clusters-and-availability/xdcr-conflict-resolution.adoc:
```diff
@@ -5,45 +5,30 @@
 . Expiration (TTL) value
 . Document flags
 
+When Cross Cluster Versioning is enabled, the Hybrid Logical Vector (HLV) metadata in the source and target documents' xattr is also used in the conflict resolution processing. For more information about the `enableCrossClusterVersioning` property and the HLV metadata, see xref:clusters-and-availability/xdcr-enable-crossclusterversioning.adoc[XDCR enableCrossClusterVersioning].
```
change xattr to xattrs

hyunjuV commented on modules/learn/pages/clusters-and-availability/xdcr-conflict-resolution.adoc:
```diff
@@ -57,6 +42,8 @@ If both document-versions have the same timestamp-value, the conflict is resolve
 . Expiration (TTL) value
 . Document flags
 
+When Cross Cluster Versioning is enabled, the Hybrid Logical Vector (HLV) metadata in the source and target documents' xattr is also used in the conflict resolution processing. For more information about the `enableCrossClusterVersioning` property and the HLV metadata, see xref:clusters-and-availability/xdcr-enable-crossclusterversioning.adoc[XDCR enableCrossClusterVersioning].
```
change xattr to xattrs

hyunjuV commented on modules/learn/pages/clusters-and-availability/xdcr-enable-crossclusterversioning.adoc:
```diff
@@ -0,0 +1,52 @@
+[#hlv-data-maintained-in-xattr]
+== Hybrid Logical Vector (HLV) data maintained in the extended attributes
+
+The new metadata, HLV, is stored as a system created extended attribute (xattr) called `_vv` (`xattrs._vv`). The HLV metadata is also called Version Vectors.
```
change attribute to attributes
change (xattr) to (xattrs)

hyunjuV commented on modules/learn/pages/clusters-and-availability/xdcr-enable-crossclusterversioning.adoc:
```diff
@@ -0,0 +1,52 @@
+[#remove-hlv-metadata]
+=== Remove HLV metadata
+
+After enabling, you cannot disable the bucket property `enableCrossClusterVersioning`. You can, however, backup and restore the data to a bucket where `enableCrossClusterVersioning` is set to `false`, and remove the xattr information added by XDCR using the option `cbbackupmgr restore --disable-hlv`. 
```
change xattr to xattrs

hyunjuV commented on modules/learn/pages/clusters-and-availability/xdcr-enable-crossclusterversioning.adoc:
```diff
@@ -0,0 +1,52 @@
+
+After enabling, you cannot disable the bucket property `enableCrossClusterVersioning`. You can, however, backup and restore the data to a bucket where `enableCrossClusterVersioning` is set to `false`, and remove the xattr information added by XDCR using the option `cbbackupmgr restore --disable-hlv`. 
+
+NOTE: To remove the on-going maintenance of HLV information, you must remove the xattr information previously added to the documents. To completely remove the effects of enabling HLV, only restoring the data to a bucket where `enableCrossClusterVersioning` is set to `false` is not enough.
```
change xattr to xattrs

hyunjuV commented on modules/rest-api/pages/rest-xdcr-adv-settings.adoc:
```diff
@@ -295,6 +292,17 @@ If successful, the call returns an object containing all current settings for th
+The following example modifies the value of `mobile` to `Active` for an existing replication:
+
+----
+curl -X POST -u Administrator:password http://localhost:8091/settings/replications/<settings_URI> -d mobile=Active | jq '.'
```
From this doc page example:
"settingsURI": "/settings/replications/2b5dcd1b0101a9d52f31a802d8c4231e%2Ftravel-sample%2Fts"

So,
<settings_URI> is /settings/replications/2b5dcd1b0101a9d52f31a802d8c4231e%2Ftravel-sample%2Fts

You could change your example either of two ways:

curl -X POST -u Administrator:password http://localhost:8091/<settings_URI> -d mobile=Active | jq '.'
Or
curl -X POST -u Administrator:password http://localhost:8091/settings/replications/<settings_URI_repl_id> -d mobile=Active | jq '.'

rao-shwe replied:
Changed to:

`curl -X POST -u Administrator:password http://localhost:8091/<settings_URI> -d mobile=Active | jq '.'`

hyunjuV commented on modules/rest-api/pages/rest-xdcr-adv-settings.adoc:
```diff
@@ -295,6 +292,17 @@ If successful, the call returns an object containing all current settings for th
 }
```
I think that it would be good to show the up-to-date 7.6.6 output for a specific replication which has the "mobile": "Off".    This means that the example output in current lines 210-243 and 259-292 need to be changed.

Note:  I am not concerned about documenting every field in the settings since many are internal settings that people should not be changing unless directed by support -- i.e. they are used in debugging/diagnostics.

So, instead of current lines 210-243 -- here it is:
{
  "casDriftThresholdSecs": 100,
  "checkpointInterval": 600,
  "ckptSvcCacheEnabled": true,
  "colMappingRules": {},
  "collectionsExplicitMapping": false,
  "collectionsMigrationMode": false,
  "collectionsMirroringMode": false,
  "collectionsOSOMode": true,
  "compressionType": "Auto",
  "dcpEnablePurgeRollback": false,
  "desiredLatency": 50,
  "docBatchSizeKb": 2048,
  "failureRestartInterval": 10,
  "filterBinary": false,
  "filterBypassExpiry": false,
  "filterBypassUncommittedTxn": false,
  "filterDeletion": false,
  "filterExpiration": false,
  "filterExpression": "",
  "jsFunctionTimeoutMs": 20000,
  "logLevel": "Info",
  "mergeFunctionMapping": {},
  "mobile": "Off",
  "networkUsageLimit": 0,
  "optimisticReplicationThreshold": 256,
  "pauseRequested": false,
  "preCheckCasDriftThresholdHours": 8760,
  "preReplicateVBMasterCheck": true,
  "priority": "High",
  "replicateCkptIntervalMin": 20,
  "retryOnErrExceptAuthErrMaxWaitSec": 360,
  "retryOnRemoteAuthErr": true,
  "retryOnRemoteAuthErrMaxWaitSec": 360,
  "skipReplSpecAutoGc": false,
  "sourceNozzlePerNode": 2,
  "statsInterval": 1000,
  "targetNozzlePerNode": 2,
  "targetTopologyLogFrequency": 1800,
  "type": "xmem",
  "workerBatchSize": 500
}

So, instead of current lines 259-292 -- here it is:


{
  "casDriftThresholdSecs": 100,
  "checkpointInterval": 700,
  "ckptSvcCacheEnabled": true,
  "colMappingRules": {},
  "collectionsExplicitMapping": false,
  "collectionsMigrationMode": false,
  "collectionsMirroringMode": false,
  "collectionsOSOMode": true,
  "compressionType": "Auto",
  "dcpEnablePurgeRollback": false,
  "desiredLatency": 50,
  "docBatchSizeKb": 2048,
  "failureRestartInterval": 10,
  "filterBinary": false,
  "filterBypassExpiry": false,
  "filterBypassUncommittedTxn": false,
  "filterDeletion": false,
  "filterExpiration": false,
  "filterExpression": "",
  "jsFunctionTimeoutMs": 20000,
  "logLevel": "Info",
  "mergeFunctionMapping": {},
  "mobile": "Off",
  "networkUsageLimit": 0,
  "optimisticReplicationThreshold": 256,
  "pauseRequested": false,
  "preCheckCasDriftThresholdHours": 8760,
  "preReplicateVBMasterCheck": true,
  "priority": "High",
  "replicateCkptIntervalMin": 20,
  "retryOnErrExceptAuthErrMaxWaitSec": 360,
  "retryOnRemoteAuthErr": true,
  "retryOnRemoteAuthErrMaxWaitSec": 360,
  "skipReplSpecAutoGc": false,
  "sourceNozzlePerNode": 2,
  "statsInterval": 1000,
  "targetNozzlePerNode": 2,
  "targetTopologyLogFrequency": 1800,
  "type": "xmem",
  "workerBatchSize": 500
}

hyunjuV reviewed (approved):
I had a few minor comments.  I trust that you'll deal with them as appropriate.

I am going to go ahead and approve so that you won't be held up.

hyunjuV commented on modules/rest-api/pages/rest-xdcr-adv-settings.adoc:
Also, on this page, for the output from 
curl -u Administrator:password -X GET http://localhost:8091/settings/replications | jq '.' 
Lines 96-123
The current output would look like:

{
  "casDriftThresholdSecs": 100,
  "checkpointInterval": 600,
  "ckptSvcCacheEnabled": true,
  "collectionsOSOMode": true,
  "compressionType": "Auto",
  "dcpEnablePurgeRollback": false,
  "desiredLatency": 50,
  "docBatchSizeKb": 2048,
  "failureRestartInterval": 10,
  "filterBinary": false,
  "filterBypassExpiry": false,
  "filterBypassUncommittedTxn": false,
  "filterDeletion": false,
  "filterExpiration": false,
  "genericServicesLogLevel": {
    < ... diagnostic items cut out due to length ... >
  },
  "goGC": 100,
  "goMaxProcs": 4,
  "jsFunctionTimeoutMs": 20000,
  "logLevel": "Info",
  "mergeFunctionMapping": {},
  "mobile": "Off",
  "networkUsageLimit": 0,
  "optimisticReplicationThreshold": 256,
  "preCheckCasDriftThresholdHours": 8760,
  "preReplicateVBMasterCheck": true,
  "priority": "High",
  "replicateCkptIntervalMin": 20,
  "retryOnErrExceptAuthErrMaxWaitSec": 360,
  "retryOnRemoteAuthErr": true,
  "retryOnRemoteAuthErrMaxWaitSec": 360,
  "skipReplSpecAutoGc": false,
  "sourceNozzlePerNode": 2,
  "statsInterval": 1000,
  "targetNozzlePerNode": 2,
  "targetTopologyLogFrequency": 1800,
  "workerBatchSize": 500
}

rao-shwe replied:
I've added (removed old content) this as-is.

rao-shwe committed:
implemented-review2-inputs-from-pm

rao-shwe committed:
minor-edits

rao-shwe committed:
minor edits for links

ggray-cb commented:
Hi @rao-shwe, I removed the password for the preview site from your comment, as the preview site and this PR are accessible by the public. Please don't expose the password in PRs. Instead, you can point people to the [Confluence page with the passwords](https://confluence.issues.couchbase.com/wiki/spaces/DOCS/pages/1971224949/Docs+Team+demo+site+passwords) in it. That page isn't publicly accessible.

ggray-cb commented on modules/learn/pages/clusters-and-availability/xdcr-active-active-sgw.adoc:
```diff
@@ -0,0 +1,83 @@
+[#xdcr-active-active-sgw-intro]
+== Introduction
+
+In the versions earlier than Server 7.6.6 and Sync Gateway (SGW) 4.0.0, only an active-passive setup was supported with both XDCR and SGW. XDCR Active-Active replication with Sync Gateway for XDCR-Mobile interoperability configuration is introduced in the Server 7.6.6 version, where you can configure an active-active XDCR setup with Sync Gateway (SGW) and mobile applications both on the XDCR sour ...
```
You don't need to change this now, but for future reference. Our standards call for "[ventilated prose.](https://docs.couchbase.com/home/contribute/pages.html#ventilate)" Each sentence should start on its own line of the text file. Among other reasons, it makes it easier for tools that spot changes in the source code to mark what has changed and what has remained the same in a page.

rao-shwe replied:
Thanks, Gary. And I was looking for that term.. "ventilated prose".
Ray gave me the same feedback earlier today and I've implemented it.

ggray-cb commented on modules/learn/pages/clusters-and-availability/xdcr-active-active-sgw.adoc:
```diff
@@ -0,0 +1,83 @@
+====
+Here are a few limitations to the _XDCR Active-Active with Sync Gateway_ feature.
+
+* If you use the _user created extended attributes (user xattrs)_ in your documents, and you have more than 10 user xattrs in a document, then you cannot use the feature _XDCR Active-Active with Sync Gateway_. This is due to an internal limitation of managing extended attributes in a document. If you try to use the feature _XDCR Active-Active with Sync Gateway_ when you have more than 10 user xat ...
```
Our standards say [not to use italics](https://docs.couchbase.com/styleguide/italic.html). I know they are all throughout the current docs. They should be removed when you find them. Also, bold should be limited to UI elements in the Web UI. The Prometheus stat should be in monospace (`subdoc_cmd_docs_skipped`).

rao-shwe replied:
I'll follow Italics related standard from my next doc task.

ggray-cb commented on modules/learn/pages/clusters-and-availability/xdcr-active-active-sgw.adoc:
```diff
@@ -0,0 +1,83 @@
+
+To configure a new active-active XDCR with Sync Gateway setup, do the following:
+
+. Create two clusters on Server 7.6.6 or a higher version with _all_ the nodes of the clusters, for example, cluster A and cluster B (or you can upgrade the existing Server clusters to 7.6.6 or a higher version).
```
I would break this bullrt point into two sentences at the "for example. " It's a bit long otherwise.

ggray-cb commented on modules/learn/pages/clusters-and-availability/xdcr-active-active-sgw.adoc:
```diff
@@ -0,0 +1,83 @@
+[#xdcr-active-active-sgw-upgrade]
+== Upgrading an existing setup
+
+Convert an existing active-passive XDCR-Sync Gateway (SGW) setup into an active-active XDCR-Sync Gateway setup.
```
I would add "You can..." to the start of this sentenc. Otherwise, it sounds like the user **must** do this, rather than this being an option. Unless this is now required, then saying "You must..." would be appropriate.

RayOffiah commented on modules/introduction/partials/new_features-76_6.adoc:
```diff
@@ -0,0 +1,5 @@
+=== XDCR
+
+* *https://jira.issues.couchbase.com/browse/MB-57921[MB-57921]:*
+Created provision to set up XDCR bi-directional replication with Sync Gateway (SGW) 4.0+. In the versions earlier than Server 7.6.6 and Sync Gateway (SGW) 4.0.0, only an active-passive setup was supported with both XDCR and SGW. XDCR active-active replication with Sync Gateway for XDCR-Mobile interoperability configuration is introduced in the Server 7.6.6 version, where you can configure an acti ...
```
We usually have each sentence on a separate line. I think there is a name for it, but I can't remember what we call it.

rao-shwe replied:
I've made this change in all the pages/sections I've worked on for this PR.

RayOffiah commented on modules/learn/pages/clusters-and-availability/xdcr-active-active-sgw.adoc:
```diff
@@ -0,0 +1,83 @@
+====
+Here are a few limitations to the _XDCR Active-Active with Sync Gateway_ feature.
+
+* If you use the _user created extended attributes (user xattrs)_ in your documents, and you have more than 10 user xattrs in a document, then you cannot use the feature _XDCR Active-Active with Sync Gateway_. This is due to an internal limitation of managing extended attributes in a document. If you try to use the feature _XDCR Active-Active with Sync Gateway_ when you have more than 10 user xat ...
```
Yeah, one sentence per-line makes it easier to flag changes.

RayOffiah replied:
```suggestion
* If you use the _user created extended attributes (user xattrs)_ in your documents, and you have more than 10 user xattrs in a document, then you cannot use the feature _XDCR Active-Active with Sync Gateway_. This is due to an internal limitation of managing extended attributes in a document. If you try to use the feature _XDCR Active-Active with Sync Gateway_ when you have more than 10 user xattrs in your document, the XDCR replication **silently skips** replicating that document. As a result, the data in the replication-skipped document will not be consistent between the target and source clusters. The only way you will know this skip occured is because the Prometheus stat **subdoc_cmd_docs_skipped** will be incremented and the document will _not_ be consistent between the target and source.
```

RayOffiah commented on modules/learn/pages/clusters-and-availability/xdcr-conflict-resolution.adoc:
```diff
@@ -98,8 +85,15 @@ When availability is restored to data center A, applications must wait for the s
+
+[IMPORTANT]
+====
+* You must select the same conflict resolution policy for all the buckets in the replication topology because you can create a replication between only those buckets that have the same conflict resolution policy. 
```
```suggestion
* You must select the same conflict resolution policy for all the buckets in the replication topology. 
```

rao-shwe replied:
We need this full sentence. So, I'm keeping it as-is.

RayOffiah commented on modules/learn/pages/clusters-and-availability/xdcr-conflict-resolution.adoc:
```diff
@@ -98,8 +85,15 @@ When availability is restored to data center A, applications must wait for the s
+[IMPORTANT]
+====
+* You must select the same conflict resolution policy for all the buckets in the replication topology because you can create a replication between only those buckets that have the same conflict resolution policy. 
+* When creating a bucket, you must actively choose the conflict resolution policy and if you do not choose a policy, the Sequence number-based conflict resolution policy is set as default. 
```
```suggestion
* When creating a bucket, you must actively choose the conflict resolution policy. If you do not choose a policy, the Sequence number-based conflict resolution policy is set as default.```

RayOffiah commented on modules/learn/pages/clusters-and-availability/xdcr-conflict-resolution.adoc:
```diff
@@ -114,9 +108,9 @@ Therefore, in this instance, timestamp-based conflict resolution should be used,
+XDCR replications cannot be created between buckets with different conflict resolution policies. The source and target buckets must always be configured with the same conflict resolution policy.
 
-When using XDCR with a source cluster running a pre-4.6 version of Couchbase Server, only conflict resolution based on _sequence numbers_ can be used.
+When creating a bucket, you must actively choose the conflict resolution policy and if you do not choose a policy, the Sequence number-based conflict resolution policy is set as default. After the bucket is created, you cannot change the conflict resolution policy for that bucket. In general, the Timestamp-based conflict resolution policy is preferred as the logic is easier to understand.
```
```suggestion
When creating a bucket, you must actively choose the conflict resolution policy. 
If you do not choose a policy, the Sequence number-based conflict resolution policy is set as default. After the bucket is created, you cannot change the conflict resolution policy for that bucket. In general, the Timestamp-based conflict resolution policy is preferred as the logic is easier to understand.
```

RayOffiah commented on modules/learn/pages/clusters-and-availability/xdcr-enable-crossclusterversioning.adoc:
```diff
@@ -0,0 +1,52 @@
+[#modify-enablecrossclusterversioning]
+== Modify the bucket property enableCrossClusterVersioning
+
+You can modify the bucket property `enableCrossClusterVersioning` through REST API. For information about modifying `enableCrossClusterVersioning` through REST API, see xref:rest-api:rest-bucket-create.adoc#example-enablecrossclusterversioning-edit[Example: Turning on enableCrossClusterVersioning, when Editing].
```
```suggestion
You can modify the bucket property `enableCrossClusterVersioning` through the  REST API. For information about modifying `enableCrossClusterVersioning` through REST API, see xref:rest-api:rest-bucket-create.adoc#example-enablecrossclusterversioning-edit[Example: Turning on enableCrossClusterVersioning, when Editing].
```

RayOffiah commented on modules/manage/pages/manage-xdcr/create-xdcr-replication.adoc:
```diff
@@ -46,7 +46,7 @@ These are named after their IP addresses: `10.144.210.101` and `10.144.210.102`.
 == Create an XDCR Replication with the UI
 
-Proceed as follows:
+Proceed as follows: 
```
```suggestion
Proceed as follows:
```

RayOffiah commented on modules/rest-api/pages/rest-bucket-create.adoc:
```diff
@@ -19,7 +19,7 @@ POST /pools/default/buckets/<bucketName>
 
 These respectively create a new bucket and edit an existing bucket.
-The bucket can be of any type: Couchbase, Ephemeral, or Memcached.
+The bucket can be of any type: Couchbase, Ephemeral, or Memcached. 
```
Watch out for odd spaces being added to the end of lines. It causes false changes to show up in the version control.

RayOffiah reviewed (changes requested):
Just a few minor points. 
We usually put one sentence on a line on its own, but I wouldn't change them now.

rao-shwe committed:
editorial-review1a-implemented

rao-shwe committed:
editorial-review-1b-implemented

rao-shwe commented:
@RayOffiah,
I implemented your review comments in VS Code in one go. So, I did not use the inline 'comment resolution' option here on GitHub.

ggray-cb reviewed (approved):
Left some comments, either for later or just to make note of for your next project.

sumukhbhat2701 commented on modules/learn/pages/clusters-and-availability/xdcr-active-active-sgw.adoc:
```diff
@@ -0,0 +1,94 @@
+image::clusters-and-availability/xdcr-active-sgw-before-upgrade.png[,720,align=left]
+
+. Upgrade both clusters A and B with _all_ the nodes of the clusters to Server 7.6.6 or a higher version.
+. Enable ECCV on bucket B1. All the mutations in B1, after this point of time, will have a new metadata called HLV.
```
After this sentence can we add a note of caution in brackets saying that - 
```
(Note that if there are more than two buckets in the replication topology, one must enable ECCV for ALL the buckets in that topology)
```

sumukhbhat2701 commented on modules/learn/pages/clusters-and-availability/xdcr-active-active-sgw.adoc:
```diff
@@ -0,0 +1,94 @@
++
+NOTE: ECCV refers to the bucket property `enableCrossClusterVersioning`. 
++
+. Enable ECCV on bucket B2. All the mutations in B2, after this point of time, will have a new metadata called HLV.
```
After this sentence can we add a note of caution saying that - 
```
NOTE: Note that if there are more than two buckets in the replication topology, one must enable ECCV for ALL the buckets in that topology
```

sumukhbhat2701 replied:
Maybe we can move the above note of:
```
NOTE: ECCV refers to the bucket property `enableCrossClusterVersioning`.
```
here below and combine as:
```
NOTE: ECCV refers to the bucket property `enableCrossClusterVersioning`. If there are more than two buckets in the replication topology, one must enable ECCV for ALL the buckets in that topology.
```

sumukhbhat2701 commented on modules/learn/pages/clusters-and-availability/xdcr-active-active-sgw.adoc:
```diff
@@ -0,0 +1,94 @@
+
+. Create two clusters on Server 7.6.6 or a higher version with _all_ the nodes of the clusters, for example, cluster A and cluster B (or you can upgrade the existing Server clusters to 7.6.6 or a higher version).
+. Create buckets, for example, B1 and B2 in cluster A and cluster B respectively, between which XDCR will be set up. Now, do the following:
+.. Enable the ECCV setting on B1. All the mutations in B1 will have a new metadata called HLV.
```
Same applies for both the sentences here - https://github.com/couchbase/docs-server/pull/3801/files#r2079167984

sumukhbhat2701 reviewed (changes requested):
@rao-shwe, sorry for the late comments, but I realised something today for readers to not make a mistake and I hope I've caught the train on time.

sumukhbhat2701 commented on modules/learn/pages/clusters-and-availability/xdcr-active-active-sgw.adoc:
```diff
@@ -0,0 +1,94 @@
+.. Enable the ECCV setting on B1. All the mutations in B1 will have a new metadata called HLV.
+.. Enable the ECCV setting on B2. All the mutations in B2 will have a new metadata called HLV.
++
+NOTE: ECCV refers to the bucket property `enableCrossClusterVersioning`.
```
Same here - https://github.com/couchbase/docs-server/pull/3801/files/3351bddd1b4000904a51784bec8d3ffb66f72904#r2079168663

rao-shwe committed:
eccv-multiple-buckets-note-added

rao-shwe commented:
@sumukhbhat2701 
I've implemented your comments. Could you check?
I've added NOTE for ECCV related to multiple buckets in 3 different places: lines 51, 77, and 81 of the file modules/learn/pages/clusters-and-availability/xdcr-active-active-sgw.adoc

rao-shwe commented:
> @rao-shwe, sorry for the late comments, but I realised something today for readers to not make a mistake and I hope I've caught the train on time.

No worries!

rao-shwe committed:
highlighted-sgw-4-as-a-future-release

rao-shwe committed:
minor-edit

rao-shwe committed:
added-sgw-4-version-note

RayOffiah committed:
Merge branch 'release/7.6' into DOC_12667_xdcr_mobile_coexist_766
